chore: CustomerIOPluginOptionsIOS structure revised for clarity

Ahmed-Ali opened the pull request:
TL;DR:
* All push relevant configuration moved from being properties of `CustomerIOPluginOptionsIOS` to be properties of `CustomerIOPluginOptionsIOS.pushNotification`

--

The `CustomerIOPluginOptionsIOS` type declares the properties use to configure the how the expo plugin will setup CustomerIO SDK.

Part of this is the push notification configurations which are meant to grouped together in the `pushNotification` field for clarity. However, we found out that some push notification configuration flags are placed as fields of the top level type. Moreover, due to this misplacement, some parts of our code expected these flags to be set under the `pushNotification` properties while they aren't and that introduced bugs.

This PR revising this structure to make the object structure easier to reason about.

The impacted properties will be deprecated and will be removed by end of Q2.

## Change Log
In the app.json and app.config.js all the following push notification flags under the `ios`

```json
{
  "ios": {
    "showPushAppInForeground": true,
    "autoTrackPushEvents": true,
    "handleDeeplinkInKilledState": true,
    "disableNotificationRegistration": true,
    "autoFetchDeviceToken": true
  }
}

```
Are move to be part of the `pushNotification` object instead
```json
{
  "ios": {
    "pushNotification": {
      "showPushAppInForeground": true,
      "autoTrackPushEvents": true,
      "handleDeeplinkInKilledState": true,
      "disableNotificationRegistration": true,
      "autoFetchDeviceToken": true
    }
  }
}

```

uros-mil commented on plugin/src/ios/withNotificationsXcodeProject.ts:
```diff
@@ -299,24 +293,24 @@ const updateNseEnv = (
 
```
Let's add early-exit if `options.env` is not object and doesn't exist.
This will make rest of the code easier to read.

We could also deconstruct env to directly use `cdPApiKey` and `region`.

Ahmed-Ali replied:
I am not sure where the options.env is used in this snippet?
In fact most of the changes here seems to be either: formatting that was missed earlier or my local eslint is kicking in and adding more formatting that we intend.

Shahroz16 commented on plugin/src/ios/withAppDelegateModifications.ts:
```diff
@@ -240,19 +240,13 @@ export const withAppDelegateModifications: ConfigPlugin<
-        props.disableNotificationRegistration !== undefined &&
-        props.disableNotificationRegistration === false
-      ) {
+      if (props.pushNotification?.disableNotificationRegistration === false) {
```
+1 to what Uros said, the previous and new check is not the same as previous one. Having said that 

```
props.pushNotification?.disableNotificationRegistration !== true
```

would be different too, 

```
In the original version, the condition is only true when the value exists and is explicitly false.
In the PR version, the condition is only true when the value is explicitly false.
In the uros version, the condition is true for any value except true. This includes undefined, null, false, 0, empty strings, etc.
```

Ahmed-Ali replied:
The PR implementation matches exactly the current production implementation and the description on top of the condition.

Seeing these comments makes me realize that this condition is currently a production bug. Changing `===false` to `!== true` should fix it as Uros suggested. I will update the comment too as it is currently incorrect as well.

mrehan27 commented on plugin/src/ios/withCIOIos.ts:
```diff
@@ -1,21 +1,86 @@
+
+  if (props.pushNotification?.env !== undefined) {
+    console.warn(
+      'The ios.pushNotification.env property is deprecated. Use ios.nv instead.'
```
```suggestion
      'The ios.pushNotification.env property is deprecated. Use ios.env instead.'
```

mrehan27 commented on plugin/src/ios/withNotificationsXcodeProject.ts:
```diff
@@ -299,24 +293,24 @@ const updateNseEnv = (
   let envFileContent = FileManagement.readFile(envFileName);
 
-  if (options.pushNotification?.env?.cdpApiKey) {
+  if (options.env?.cdpApiKey) {
```
At the moment, `env` values are only used for NSE when rich push notifications are enabled on iOS. That's why I feel having `env` under `ios.pushNotification` was slightly self explanatory in terms of its purpose. Just wanted to double-check, knowing this, do we still want to move `env` outside of `pushNotification` and directly under `ios` instead? 🤔

Also, just wanted to confirm my understanding. The `mergeDeprecatedPropertiesAndLogWarnings` function above should update properties to always match the new format. Since that always runs first, we only need to read props from new location everywhere else, Correct?

Ahmed-Ali replied:
Ah good point! In that case, they do make sense to be inside push. There is an argument to be made about their best placement, but no strong opinion at this point from my side, hence I will revert that change

mrehan27 commented on plugin/src/ios/withCIOIos.ts:
```diff
@@ -1,21 +1,86 @@
+const mergeDeprecatedPropertiesAndLogWarnings = (
+  props: CustomerIOPluginOptionsIOS
+) => {
+  const deprecatedTopLevelProperties = {
```
Not directly related to this code, but adding comment here so we don't forget. Would be nice if we can update compatibility scripts to use newer properties as well (e.g. [`env` here](https://github.com/customerio/customerio-expo-plugin/blob/37415dcc8081cc2a886f347a763f7310ad2a9295/scripts/compatibility/configure-plugin.js#L107-L110)).

Ahmed-Ali replied:
That makes a lot of sense.
I will add the comments on this one. I also checked the `env` file and since I will revert the `props.pushNotification.env` change, I didn't see anything else there that will be impacted.
I will ping you after I address the rest of the comment to double check with me in case I missed something.

mrehan27 reviewed (commented):
The compatibility tests are failing. They likely just need script updates to add config properties at new location. That should fix it and we should be good to go.

Ahmed-Ali commented on plugin/src/ios/withCIOIos.ts:
```diff
@@ -1,21 +1,86 @@
+  // loop over all the deprecated properties and log a warning if they are set
+  Object.entries(deprecatedTopLevelProperties).forEach(([key, value]) => {
+    if (value !== undefined) {
+      console.warn(
```
I don't think this is the intention here.
These fields today are at the top level and we are deprecating them in favor of moving them to the `pushNotification` sub object. Hence the warning makes sense to be here. The second warning will only happen if the user set the value in the correct place and in the deprecated field. Which is a second case and its warning seems to also make sense.

Ahmed-Ali commented on plugin/src/ios/withNotificationsXcodeProject.ts:
```diff
@@ -30,7 +30,7 @@ const addNotificationServiceExtension = async (
     }
 
-    if (options.pushNotification?.useRichPush) {
+    if (options.pushNotification?.useRichPush === true) {
```
This matches our current [documented defaults](https://docs.customer.io/sdk/expo/getting-started/packages-options/#:~:text=ios.pushNotification.useRichPush) better

mrehan27 reviewed (commented):
Changes look good. Just two things before we merge:

- We need to update base to `main` since we replaced `master` with `main` recently
- We need to update snapshot tests to ensure CI checks are passing

Rest looks good 👍🏻

mahmoud-elmorabea commented on __tests__/ios/apn/AppDelegate-impl.test.js:
```diff
@@ -21,7 +21,7 @@ test('Plugin injects CIO imports and calls into AppDelegate.mm', async () => {
     // Add swift bridge imports
     #import <ExpoModulesCore-Swift.h>
-    #import <${testAppName()}-Swift.h>
+    #import <ExpoTestbed-Swift.h>
```
This should remain dynamic so it works for the base test app as well as the compatibility apps we run tests against different versions of Expo, any reason to change this?

mahmoud-elmorabea commented on __tests__/ios/fcm/AppDelegate-impl.test.js:
```diff
@@ -20,11 +20,10 @@ test('Plugin injects CIO imports and calls into AppDelegate.mm', async () => {
     // Add swift bridge imports
     #import <ExpoModulesCore-Swift.h>
-    #import <${testAppName()}-Swift.h>
+    #import <ExpoTestbed-Swift.h>
```
This should remain dynamic so it works for the base test app as well as the compatibility apps we run tests against different versions of Expo, any reason to change this?

Ahmed-Ali replied:
For a second I was like "I never touched these files". It took place when I tried to get jest to help update the snapshot tests `npm run jest -u`. Good learning that we might want to put a hook to fix it on commit

mahmoud-elmorabea commented on __tests__/ios/fcm/__snapshots__/NotificationService-swift.test.js.snap:
```diff
@@ -3,7 +3,7 @@
 "import Foundation
 import UserNotifications
-import CioMessagingPushFCM
+import CioMessagingPushAPN
```
Looking at the path of the test, this test is specific to FCM and it should import the FCM messaging package

mahmoud-elmorabea commented on __tests__/ios/fcm/__snapshots__/NotificationService-swift.test.js.snap:
```diff
@@ -12,7 +12,7 @@ public class NotificationServiceCioManager : NSObject {
   @objc(didReceive:withContentHandler:)
   public func didReceive(_ request: UNNotificationRequest, withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void) {
-    MessagingPushFCM.initializeForExtension(
+    MessagingPushAPN.initializeForExtension(
```
Looking at the path of the test, this test is specific to FCM and it should import the FCM messaging package

Ahmed-Ali commented:
Thanks Rehan for the review. The PR is now rebased on top of base. I updated the snapshot tests that's currently working. There are few others who are currently broken in production and worth addressing on their own right rather than trying to squeeze them with this change

mahmoud-elmorabea commented on plugin/src/ios/withCIOIos.ts:
```diff
@@ -10,12 +10,63 @@ export function withCIOIos(
+  // still available for backwards compatibility, but they will
+  // be removed in the future.
+
+  const deprecatedTopLevelProperties = {
```
Brilliant! 💯 🥇

mahmoud-elmorabea reviewed (commented):
The core change looks great! Absolutely fantastic job @Ahmed-Ali ! I love how we're being considerate by keeping the current behavior and also giving users warnings about deprecation 👍 

One note about the broken tests, I see that the last commit to `main` branch is successfuly, why do we think the tests are broken for production?

I think few minor tweaks to the tests can fix them, I'm okay doing it in this PR or a follow up one 😇

Ahmed-Ali commented:
> One note about the broken tests, I see that the last commit to main branch is successfuly, why do we think the tests are broken for production?

When I switched to main locally FCM tests kept failing in the same way they fail on this PR. The changes you noted could be the culprit that they were still in place when I checked out main and I didn't pay attention. Will give it another go.
